tools/elasticsearch/handler.go

```diff
@@ -0,0 +1,223 @@
+	if err != nil {
+		logger.Error("Unable to load embedded index template.", tag.Error(err))
+		return err
+	}
+
+	task := SetupTask{
+		esClient: client,
+		logger:   logger,
+		config: &SetupConfig{
+			SettingsContent: settingsContent,
+			TemplateContent: templateContent,
+			VisibilityIndex: "", // Don't create index in setup-schema
+			FailSilently:    cli.Bool(CLIOptFailSilently),
+		},
+	}
+
+	return task.RunSchemaSetup()
+}
+
+// ping the elasticsearch host and return the json response
+func ping(cli *cli.Context, logger log.Logger) error {
+	client, err := createClient(cli, logger)
+	if err != nil {
+		logger.Error("Unable to read config.", tag.Error(commonschema.NewConfigError(err.Error())))
+		return err
+	}
+
+	err = client.Ping(context.TODO())
+	if err != nil {
+		logger.Error("Ping failed", tag.Error(err))
+		return err
+	}
+
+	logger.Info("Pong - Elasticsearch is reachable")
+	return nil
+}
+
+func parseElasticConfig(cli *cli.Context) (*esclient.Config, error) {
+	cfg := new(esclient.Config)
+
+	u, err := url.Parse(cli.String(CLIOptESURL))
+	if err != nil {
+		return nil, fmt.Errorf("invalid elasticsearch URL %q: %w", cli.String(CLIOptESURL), err)
+	}
+
+	cfg.URL = *u
+	cfg.Username = cli.String(CLIOptESUsername)
+	cfg.Password = cli.String(CLIOptESPassword)
+	cfg.Version = "v7" // Fixed schema version 7
+	cfg.Indices = map[string]string{}
+
+	if cli.String(CLIOptVisibilityIndex) != "" {
+		cfg.Indices[esclient.VisibilityAppName] = cli.String(CLIOptVisibilityIndex)
+	}
+
+	if cli.String(CLIOptAWSCredentials) != "" {
+		cfg.AWSRequestSigning.CredentialProvider = cli.String(CLIOptAWSCredentials)
+		cfg.AWSRequestSigning.Enabled = true
+
+		if cfg.AWSRequestSigning.CredentialProvider == "static" {
+			cfg.AWSRequestSigning.Static.AccessKeyID = cfg.Username
+			cfg.AWSRequestSigning.Static.SecretAccessKey = cfg.Password
+			cfg.AWSRequestSigning.Static.Token = cli.String(CLIOptAWSToken)
+		}
+	}
+
+	return cfg, nil
+}
+
+// updateSchema updates the index template to the latest version, or index mappings if --index is specified
+func updateSchema(cli *cli.Context, logger log.Logger) error {
+	client, err := createClient(cli, logger)
+	if err != nil {
+		logger.Error("Unable to read config.", tag.Error(commonschema.NewConfigError(err.Error())))
+		return err
+	}
+
+	templateContent, err := schema.ElasticsearchIndexTemplate()
+	if err != nil {
+		logger.Error("Unable to load embedded index template.", tag.Error(err))
+		return err
+	}
+
+	indexName := cli.String(CLIOptVisibilityIndex)
+
+	task := SetupTask{
+		esClient: client,
+		logger:   logger,
+		config: &SetupConfig{
+			SettingsContent: "", // Don't update cluster settings
+			TemplateContent: templateContent,
+			VisibilityIndex: indexName,
+			FailSilently:    cli.Bool(CLIOptFailSilently),
+		},
+	}
+
+	// If index is specified, update index mappings; otherwise update template
+	if indexName != "" {
+		return task.RunIndexUpdate()
+	}
+	return task.RunTemplateUpgrade()
```

Comment:
I think we should do both, update the template and the index. Otherwise, creating new indices will not have the right mappings.

Also, based on the `README` you wrote, I think you meant to always upgrade the template as well.

```suggestion
	if err := task.RunTemplateUpgrade(); err != nil {
		return err
	}

	// If index is specified, update index mappings; otherwise update template
	if indexName != "" {
		return task.RunIndexUpdate()
	}
```

Reply:
Yes, README behaviour was the intention, fixed now.